Notes for field WG meeting 2024-08-07 (#8345)
dotnet/csharplang: meetings/working-groups/field-keyword/FK-2024-08-07 Nullability analysis with the `field` keyword.md

# Nullability analysis with the `field` keyword

Table of contents:

- [Motivation](#motivation)
- [Introduction](#introduction)
- [Constructor initialization](#constructor-initialization)
- [Field-targeted nullability attributes](#field-targeted-nullability-attributes)
- [Property-targeted nullability attributes](#property-targeted-nullability-attributes)
- [Null-resilient getters](#null-resilient-getters)
- [Alternatives](#alternatives)

## Motivation

Some vastly common patterns which are served by the `field` feature are lazy initialization, and defaulting. When a field-backed property is non-nullable, it's required to be initialized during construction. This flies in the face of these kinds of properties. They are well-formed. Warnings requiring constructor initialization would be noisy and unwelcome in these properties and in many similar forms.

```cs
public List<int> LazyProp => field ??= new();

public string DefaultingProp { get => field ?? GetDefault(); private set; }
```

These scenarios are made possible using the existing nullability attributes, like `[MaybeNull]`, just as they would be for manually-declared fields.

> We'll walk first through enabling the user to provide these nullability attributes explicitly for the backing field, and then through how the compiler builds on this capability to apply these attributes automatically in a simple and intuitive fashion.

Reviewer: i would like to add a comment that these use cases are not speculative.  They were a large part of the examples and conversation around `field` (and feel free to link to the discussion page there).  

Basically, i want to sidestep people going: this isn't a big deal, we don't need to do anything here.

rather, it's a very big deal.  People have been expecting these scenarios to work as part of the entire 5+ year journey this feature has been on.  Not having it would be a bitter pill.

## Introduction

This proposal brings together a few building blocks:

1. Properties can be considered initialized to non-null in the constructor by calling the setter, just as with `required` and normal flow analysis today.

1. `field`-targeted nullability attributes work just as with manually declared fields. (E.g. `[field: MaybeNull]`)

1. Null-resilient getters, a simple concept that allows maintaining a property's contract of not returning null, while flexibly allowing idiomatic patterns that use a nullable backing field. 

The following sections will walk through how these concepts play out.

## Constructor initialization

> In C# today, a non-nullable auto property is required to be initialized in the constructor. We will preserve this requirement for all field-backed properties. This requirement is in the form of the existing warning "CS8618: Non-nullable property 'Prop' must contain a non-null value when exiting constructor."

Reviewer 2: ```suggestion
In C# today, a non-nullable auto property that is not marked `required` is forced to be initialized in the constructor. We will preserve this enforcement for all field-backed properties. This enforcement is in the form of the existing warning "CS8618: Non-nullable property 'Prop' must contain a non-null value when exiting constructor."
```

Author: At this level of detail, I should probably also add "that has no property initializer and..."

Author: ```suggestion
In C# today, a non-nullable auto property which has no property initializer and is not marked `required` is forced to be initialized in the constructor. We will preserve this enforcement for all field-backed properties. This enforcement is in the form of the existing warning "CS8618: Non-nullable property 'Prop' must contain a non-null value when exiting constructor."
```

When applied, they affect the nullability analysis of the `field` keyword within the accessors just as they would affect a manually declared field. In cases where the manually-declared field would be required to be assigned in constructors, the field-backed property will be required to be assigned in constructors.

### Postconditions (MaybeNull, NotNull)

When `[field: MaybeNull]` is applied on a non-nullable field-backed property, the property will not be required to be initialized in all construction paths. This is consistent with how a manually declared non-nullable field would no longer be required to be initialized when this attribute is applied to it.

The reverse is true for `[field: NotNull]` on a *nullable* property. Just like how this would cause warnings requiring a manually declared field to be initialized in all construction paths, applying `[field: NotNull]` to a field-backed property will cause the property to be required to be initialized in all construction paths.

#### Examples 

> This will give "CS8603: Possible null reference return," on `field` just as with a manually declared field:

Reviewer 2: ```suggestion
This will give "CS8603: Possible null reference return" on `field` just as with a manually declared field:
```

                            ~~~~~
```

The same warning will be shown here on 'get', since `get;` expands to `get => field;`:

```cs
// CS8603: Possible null reference return
[field: MaybeNull]
public string Prop { get; set => field = value; }
                     ~~~
```

The next example is not real code we expect people to write, but for consistency its behavior will change. In C# 12, the attribute is ignored. In C# 13, applying this attribute to an auto-property will show the same warning as in the previous example, instead of showing "CS8618: Non-nullable property 'Prop' must contain a non-null value when exiting constructor."

```cs
// CS8603: Possible null reference return
[field: MaybeNull]
public string Prop { get; set; }
                     ~~~
```

### Preconditions (AllowNull, DisallowNull)

When `[field: AllowNull]` is applied on a non-nullable field-backed property, the property will be able to assign maybe-null values to the field without a warning. Here's an example:

```cs
[field: AllowNull]
public string ResetIfSetToDefault
{
    get => field ?? GetDefault();
    set => field = (value == GetDefault() ? null : value);
}
```

> Automatic `field` nullability is granted by the [null-resilient getter proposal](#null-resilient-getters) below. `[field: AllowNull, MaybeNull]` will be automatically applied due to the `?? GetDefault();` in the example above, and will not be applied if the `?? Default();` is removed.

Reviewer 2: ```suggestion
Automatic `field` nullability is covered by the [null-resilient getter proposal](#null-resilient-getters) below. `[field: AllowNull, MaybeNull]` will be automatically applied due to the `?? GetDefault();` in the example above, and will not be applied if the `?? Default();` is removed.
```

```cs
[AllowNull]
public string Prop { get => field ?? GetDefault(); set; }
```

Or if that version is not taken, the user will have to add `[field: AllowNull]` as well:

```cs
[AllowNull]
[field: AllowNull]
public string Prop { get => field ?? GetDefault(); set; }
```

## Null-resilient getters

A central scenario for the `field` keyword is lazily-initialized properties. Without automatic nullability, users would have to manually place `[field: MaybeNull]` on every lazily-initialized property in order for the compiler to stop telling the user that the _lazily_-initialized property should be initialized in the _constructor_!

```cs
// This is ungainly.
[field: MaybeNull]
public List<int> Prop => field ??= new();
```

A null-resilient getter is a getter which continues to fulfill the property's contract of not returning a `null` value, even when `field` is maybe-null at the start of the getter. For properties with a null-resilient getter, the backing field does not need to be initialized, and it may be assigned a maybe-null value at any point, all without risk of returning `null` from a non-nullable property.

> Null resilience is further extended to mean that there are no nullability warnings when `field` is maybe-null. This is inclusive of checking that no exit point returns a maybe-null expression. We would determine null resilience by analyzing the getter with a pass that starts `field` out as maybe-null and checks to see that there are no nullability warnings. (The less inclusive alternative is considered in [Definition of null resilience](#definition-of-null-resilience). This alternative would only check that there is no exit point returning a maybe-null expression.)

Reviewer 2: ```suggestion
Null resilience is further extended to mean that there are no nullability warnings when `field` is maybe-null. This is inclusive of checking that no exit point returns a maybe-null expression. We would determine null resilience by analyzing the getter with a pass that starts `field` out as maybe-null and checks to see that there are no nullability warnings. (A less inclusive alternative is considered in [Definition of null resilience](#definition-of-null-resilience).)
```
